### PR TITLE
GH-40079: [CI][Packaging] Enable Azure in more tests and builds

### LINKED ISSUE #40079: [CI][Packaging] Enable Azure is conda builds
### Describe the enhancement requested

Child issue of: https://github.com/apache/arrow/issues/39317

There are more builds and tests where I should have enabled Azure in https://github.com/apache/arrow/issues/39352. 

Importantly we need Azure to be enabled in python CI tests. That means enabling Azure in the conda builds. 

### Component(s)

Continuous Integration

### PR DESCRIPTION
**Pull request by Tom-Newton**
### Rationale for this change
We want python side tests of `AzureFileSystem` to run in CI. 

### What changes are included in this PR?
- Add missing `export` to enable Azure pyarrow tests
- Enable azure in sdist tests.
- Enable Azure on macos python builds
- Enable azure in conda builds and install dependencies (Azure C++ SDK and azurite)
- Enable retries on C++ tests to mitigate https://github.com/apache/arrow/issues/40121

Probably all of this should have been included in https://github.com/apache/arrow/pull/39971

### Are these changes tested?
There is no new functionality to test

### Are there any user-facing changes?
No

* Closes: #40079
* GitHub Issue: #40079

### CODE REVIEW

**Comment by Tom-Newton** [comment]
Most of the CI failures are actually reproducible on `main` with `PYTHON=3.9 docker-compose build conda-python`.

I will need to investigate the AVX2 build failure.

File: ci/docker/conda-cpp.dockerfile [hunk 1]
```diff
@@ -22,6 +22,16 @@ FROM ${repo}:${arch}-conda
 COPY ci/scripts/install_minio.sh /arrow/ci/scripts
 RUN /arrow/ci/scripts/install_minio.sh latest /opt/conda
 
+# Azurite requires npm
+RUN export DEBIAN_FRONTEND=noninteractive && \
+    apt-get update -y -q && \
+    apt-get install -y -q npm \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
```
Comment (kou) [inline ci/docker/conda-cpp.dockerfile:30]: Can we install npm (Node.js) by conda?

**Comment by kou** [comment]
Test is `AzureFileSsstem.InitializeWithDefaultCredential`:

https://github.com/apache/arrow/actions/runs/7902101938/job/21566998998?pr=40080#step:6:6540

```text
[ RUN      ] AzureFileSystem.InitializeWithDefaultCredential
```

https://github.com/apache/arrow/blob/2422994de04cf4f5a989fec0f00fabccad15b03f/cpp/src/arrow/filesystem/azurefs_test.cc#L286-L291

Important backtrace:

https://github.com/apache/arrow/actions/runs/7902101938/job/21566998998?pr=40080#step:6:6568

```text
Thread 1 (Thread 0x7f5c8e6d2cc0 (LWP 26205)):
#0  0x00007f5c92ddae4c in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (this=0x7f5c45596010, __str=...) at /home/conda/feedstock_root/build_artifacts/gcc_compilers_1706816862910/work/build/x86_64-conda-linux-gnu/libstdc++-v3/include/bits/basic_string.h:541
#1  0x000055e62c73586c in std::_Construct<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&> () at /opt/conda/envs/arrow/x86_64-conda-linux-gnu/include/c++/12.3.0/bits/stl_construct.h:119
#2  0x000055e62c7292c5 in std::__do_uninit_copy<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*> () at /opt/conda/envs/arrow/x86_64-conda-linux-gnu/include/c++/12.3.0/bits/stl_uninitialized.h:120
#3  0x00007f5c91da6f70 in Azure::Storage::_internal::StorageBearerTokenAuthenticationPolicy::Clone() const () from /opt/conda/envs/arrow/lib/./libazure-storage-common.so
#4  0x00007f5c9307cfc9 in Azure::Storage::Blobs::BlobServiceClient::BlobServiceClient(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<Azure::Core::Credentials::TokenCredential>, Azure::Storage::Blobs::BlobClientOptions const&) () from /opt/conda/envs/arrow/lib/libazure-storage-blobs.so
#5  0x00007f5c95141a35 in std::make_unique<Azure::Storage::Blobs::BlobServiceClient, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<Azure::Core::Credentials::TokenCredential>&> () at /opt/conda/envs/arrow/x86_64-conda-linux-gnu/include/c++/12.3.0/bits/unique_ptr.h:1065
#6  0x00007f5c95126c04 in arrow::fs::AzureOptions::MakeBlobServiceClient () at /arrow/cpp/src/arrow/filesystem/azurefs.cc:184
```

https://github.com/apache/arrow/blob/2422994de04cf4f5a989fec0f00fabccad15b03f/cpp/src/arrow/filesystem/azurefs.cc#L184-L185

**Comment by Tom-Newton** [comment]
Pinning to same versions we get from vcpkg solves some of the CI failures. I think the only remaining failures are ones that also exist on `main`. Specifically installing GCS testbench on python 3.9 seems to fail.

**Comment by kou** [comment]
> Pinning to same versions we get from vcpkg solves some of the CI failures.

Does it mean that we may need to improve our Azure filesystem implementation to use more recent azure-sdk-for-cpp? (If it's needed, we should work on it as a separated task.)
Or does it mean that we have symbol conflict with azure-sdk-for-cpp installed by vcpkg and conda?

Anyway, could you add a comment why we need to pin azure-sdk-for-cpp?

> I think the only remaining failures are ones that also exist on `main`. Specifically installing GCS testbench on python 3.9 seems to fail.

Could you open an issue for it so that we can work on it as a separated task?

**Comment by Tom-Newton** [comment]
Created an issue for the python 3.9 GCS testbench issue https://github.com/apache/arrow/issues/40112

For the segfault it looks like going from `azure-core-cpp` 1.10.3 to 1.11.0 causes the problem. I cannot reproduce when using the same version built from source. I also found a couple of github issue which looks very relevant 
https://github.com/conda-forge/azure-core-cpp-feedstock/pull/10#issuecomment-1890520516
https://github.com/Azure/azure-sdk-for-cpp/issues/5322

and it looks like 1.11.0 has been marked as broken on the conda website
image
https://anaconda.org/conda-forge/azure-core-cpp/labels

It seems to me like a conda issue not an Azure SDK or an Arrow Azure filesystem thing.

**Comment by Tom-Newton** [comment]
I'm going to call this ready for review but I'm not sure if we can merge it while https://github.com/apache/arrow/issues/40112 is un-resolved.

**Comment by Tom-Newton** [comment]
Hmm.... it looks like the azure tests are slightly flaky. I just changed a comment and the new CI run has an azure failure. 
Previous successful build: https://github.com/apache/arrow/actions/runs/7945396747/job/21692015614
Failure in most recent build: https://github.com/apache/arrow/actions/runs/7949050250/job/21699789831?pr=40080
```
C++ exception with description "Connection closed before getting full response or response is less than expected. Expected response length = 254. Read until now = 231" thrown in the test body.
2024-02-18T12:50:20.039Z ada6933e-9c33-47d2-86f6-29e9aa01f713 info: BlobStorageContextMiddleware: RequestMethod=DELETE RequestURL=http://127.0.0.1/devstoreaccount1/container?restype=container RequestHeaders:{"authorization":"SharedKey devstoreaccount1:hYh+JRj5cBYqdqOyM2wB3EZizQ/s2DiIoDI0CIF2EXM=","host":"127.0.0.1:10000","user-agent":"azsdk-cpp-storage-blobs/12.10.0-beta.1 (Linux 6.2.0-1019-azure x86_64 #19~22.04.1-Ubuntu SMP Wed Jan 10 22:57:03 UTC 2024)","x-ms-client-request-id":"be6819a2-72b8-4630-8eb0-4a88e7cb3061","x-ms-date":"Sun, 18 Feb 2024 12:50:20 GMT","x-ms-version":"2022-11-02"} ClientIP=127.0.0.1 Protocol=http HTTPVersion=1.1
```
It looks like azurite returned an invalid response.

I've run this multiple times locally but so far have not been able to reproduce.

**Comment by Tom-Newton** [comment]
Retrying got the same error in a different test case on a different build.

**Comment by Tom-Newton** [comment]
I think https://github.com/apache/arrow/pull/40120 should resolve the Python 3.9 and GCS testbench problem. 

I also found an example of the azure tests flaking on `main` https://github.com/apache/arrow/actions/runs/7915689559/job/21608061673. I'm unsure whether its fair to merge this while this flakiness exists. This PR enables Azure tests in more builds so the flakes will become more frequent. 

I created a separate issue to fix this flakiness https://github.com/apache/arrow/issues/40121

File: ci/conda_env_cpp.txt [hunk 2]
```diff
@@ -16,6 +16,12 @@
 # under the License.
 
 aws-sdk-cpp=1.11.68
+# There is a problem with the 1.11.0 conda release of azure-core-cpp https://github.com/conda-forge/admin-requests/pull/911
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply ci/conda_env_cpp.txt:19]: You mean 1.11 from conda? (not 1.10) Do you remember if there was more information in the logs about the crash?

[the remark this answers is not in the document]
Comment (Tom-Newton) [reply ci/conda_env_cpp.txt:19]: https://github.com/apache/arrow/pull/40080#issuecomment-1944636898 summarises the info we have. 

It looks like 1.11.1 is now available from conda though and that works, so I can remove this version restriction.

**Comment by assignUser** [comment]
@github-actions crossbow submit -g python

**Comment by Tom-Newton** [comment]
Remaining CI failures:
s3fs-test https://github.com/apache/arrow/actions/runs/7987315113/job/21809549810?pr=40080#step:7:3930
appveyor timeout https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/49237913/messages
A couple of assertion errors in pandas nightly tests https://github.com/ursacomputing/crossbow/actions/runs/7997817187/job/21842956425#step:6:6409

File: ci/scripts/cpp_test.sh [hunk 3]
```diff
@@ -86,6 +86,7 @@ ctest \
     --label-regex unittest \
     --output-on-failure \
     --parallel ${n_jobs} \
+    --repeat until-pass:3 \
```
Comment (assignUser) [inline ci/scripts/cpp_test.sh:89]: Why is this necessary? To prevent flakeyness due to network timeout? Are we running tests against remote buckets or just azurite?

**Review by assignUser — APPROVED** [review]
I agree CI failures seem unrelated but I'll let @raulcd have the last word on that. Thanks :)

**Comment by kou** [comment]
Those failures are unrelated because they are happen on main too.

**Review by kou — APPROVED** [review]
+1
